### ISSUE TITLE
LDAP Group support not working

### ISSUE THREAD

**Issue by KlavsKlavsen** [alerta/alerta#1325]
**Issue Summary**
LDAP login works, until I set this:
LDAP_DOMAINS_GROUP: "{ 'example.org': '(&(objectClass=group)(cn=API_COE_UDV_VIEW))' }"

Then I get this error (and isnt' logged in):
```
Traceback (most recent call last):
  File "/venv/lib/python3.7/site-packages/alerta/auth/basic_ldap.py", line 89, in login
    resultTypes, results = ldap_connection.result(resultID)
  File "/venv/lib/python3.7/site-packages/ldap/ldapobject.py", line 754, in result
    resp_type, resp_data, resp_msgid = self.result2(msgid,all,timeout)
  File "/venv/lib/python3.7/site-packages/ldap/ldapobject.py", line 758, in result2
    resp_type, resp_data, resp_msgid, resp_ctrls = self.result3(msgid,all,timeout)
  File "/venv/lib/python3.7/site-packages/ldap/ldapobject.py", line 765, in result3
    resp_ctrl_classes=resp_ctrl_classes
  File "/venv/lib/python3.7/site-packages/ldap/ldapobject.py", line 772, in result4
    ldap_result = self._ldap_call(self._l.result4,msgid,all,timeout,add_ctrls,add_intermediates,add_extop)
  File "/venv/lib/python3.7/site-packages/ldap/ldapobject.py", line 338, in _ldap_call
    reraise(exc_type, exc_value, exc_traceback)
  File "/venv/lib/python3.7/site-packages/ldap/compat.py", line 46, in reraise
    raise exc_value
  File "/venv/lib/python3.7/site-packages/ldap/ldapobject.py", line 322, in _ldap_call
    result = func(*args,**kwargs)
ldap.OPERATIONS_ERROR: {'msgtype': 100, 'msgid': 2, 'result': 1, 'desc': 'Operations error', 'ctrls': [], 'info': '000004DC: LdapErr: DSID-0C090A7D, comment:
In order to perform this operation a successful bind must be completed on the connection., data 0, v3839'}

**Environment**
- API version: 8.0.3
- Deployment:Docker on k8s
- Database: Postgres

- Server config:
  Auth enabled? Yes
  Auth provider? LDAP
  Customer views? No
- web UI version: [eg. 8.0.2]

**Comment by KlavsKlavsen** [comment]
This seems to be a python-ldap problem. https://code.i-harness.com/en/q/22497
It may be that my user does not have the save privs as the user you'd normally use to bind (instead of binding with the user who is logging in). 
I think merging https://github.com/alerta/alerta/pull/1206 would make it work.
I'll try and patch that and put in my own docker image to test.

**Comment by satterly** [comment]
I think once simple bind has been used to authenticate the user that wants to log in the "bind user" should be used again to get the authenticated user attributes and groups. ie. not just use the authenticated user to get those details.

**Comment by KlavsKlavsen** [comment]
Thats probaly correct @satterly - one might imagine a situation where normal users aren't allowed to make those kinda queries

**Comment by satterly** [comment]
Please test and confirm the fix, if you can. Thanks.

**Comment by KlavsKlavsen** [comment]
@satterly Making it kinda hard to build a test docker image: https://github.com/alerta/docker-alerta/blob/master/Dockerfile#L55 - got any hints for making it easier? other than moving Dockerfile to alerta repo instead and figureing out what to manually copy into docker image? (We ONLY run things on k8s here - so I need to build a docker image to test :)

**Comment by satterly** [comment]
That's the official docker image that is only bumped when there's a new release. To deploy and test master branch use  ...

```
$ docker pull docker.pkg.github.com/alerta/alerta/alerta-api:latest
```

**Comment by KlavsKlavsen** [comment]
Appearently only you can read that ? It says to verify my personal access has the correct scopes :(

**Comment by KlavsKlavsen** [comment]
I tried updating my local fork - but I just got an email about my docker build and push failing :(
seems to be due to some slack hook not working (which it won't as I don't have that token :)

**Comment by KlavsKlavsen** [comment]
could you push latest tag to docker hub perhaps?

**Comment by satterly** [comment]
you just need to do a docker login using a github access token to get access to github package repos.

**Comment by KlavsKlavsen** [comment]
I did do a docker login with my github account and an access token.. found the reason - when you copy paste token password  it gives you a neat little space in front.. :)

**Comment by KlavsKlavsen** [comment]
is the alerta-api build here - directly able to replace the alerta-web image thats used normally by docker-alerta ?

**Comment by satterly** [comment]
No it's only the API. Sorry should have said.

**Comment by KlavsKlavsen** [comment]
Soo.. how do I best test LDAP login using only API.. :)
Perhaps I can add a pull of alerta-web and put that in the image myself.. Which Dockerfile are you using to build alerta-api? (it must be in the repo somewhere I assume) - then I'll try and modify that to take an option for pulling web or something to build a complete one for testing.

**Comment by KlavsKlavsen** [comment]
yikes. ignore my stupid question :) https://github.com/alerta/alerta/blob/master/Dockerfile -> I'll try to add web to that and build my own for testing :)

**Comment by satterly** [comment]
Or just use two images. The `alerta-web` just for the web UI and `alerta-api` for the API.

If using docker-compose it might look something like this ...
```
  webui:
    image: alerta/alerta-web
    networks:
      net:
        aliases:
          - web

  api:
    image: alerta-api
    depends_on:
      - db
    volumes:
      - ./api/alertad.conf:/app/alertad.conf
    environment:
      - DEBUG=1  # remove this line to turn DEBUG off
      - DATABASE_URL=postgres://postgres:postgres@db:5432/monitoring
      - AUTH_REQUIRED=True
      - ADMIN_USERS=admin@alerta.io,devops@alerta.io #default password: alerta
      - ADMIN_KEY=demo-key  # assigned to first user in ADMIN_USERS list
      # - PLUGINS=reject,blackout,normalise,enhance
    networks:
      net:
        aliases:
          - api
```

If you're struggling ping me on gittter https://gitter.im/alerta/chat

**Comment by KlavsKlavsen** [comment]
Thank you for the offer @satterly  - but WTF is up with that gitter.im thing? If I try to login with github - it wants access to groups - where I can't really read exactly WHAT access it gets.. Why should it ask for ANYTHING beyond my username+email ?
If I choose gitlab - its even worse:  "Access the authenticated user's API
Grants complete read/write access to the API, including all groups and projects, the container registry, and the package registry."
- no way I'm going to grant gitter full write access to my gitlab account.
and using twitter - it wants to read all my posts, blacklists etc.. again WTF

and I have ZERO options for creating my own user - just auth'ed via email as normal.

Do they not want users - or do they purposely NOT want to live up to GDPR - the above is a STRICT violation of the GDPR (EU regulation)..
as an EU citizen I cannot accept any of those options (I've posted to them about it on twitter to start with).. I ended up taking the docker-alerta Dockefile and adjust and copy in a few files to alerta repo - to build a workable docker image with the helm chart
